Deliver mails in context of their recipient

NobodysNightmare opened the pull request:
## OpenProject work package

https://community.openproject.org/work_packages/19769
## Description

This PR changes mail notification so that they set the current user to the recipient of the mail while it is delivered. Thereby it is ensured that methods like `WorkPackage.visible` work during mail creation. This is required for text formatting, e.g. to turn `##123` into `Bug 123: A bug!`

NobodysNightmare committed:
use more specific names than user

author and recipient are matching the domain they are employed in (mail delivery) closer.

Especially the name 'current_user' only happens to be the thing being assigned as author,
but does not transport any special meaning in itself.

NobodysNightmare committed:
pull recipient & author up into MailNotificationJob

NobodysNightmare committed:
execute mail jobs in context of their recipient

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://nobodysnightmare-feature-recipient-context.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://nobodysnightmare-feature-recipient-context.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

NobodysNightmare committed:
style

NobodysNightmare committed:
move error handling up

RecordNotFound would now already occur due to :recipient call in :perform method.

NobodysNightmare commented on app/workers/mail_notification_job.rb:
```diff
@@ -29,16 +29,20 @@
-    notify
+  def initialize(recipient_id, author_id)
+    @recipient_id = recipient_id
+    @author_id    = author_id
```
I do not agree and rubocop does not agree either.

Azure7111 commented:
So far this looks good. But the spec failures from the previous completed run seem legit. Lets wait what the next run of Travis finds. You may however add tests for `execute_as` in the meantime as that method has probably no coverage right now.

ulferts commented:
@NobodysNightmare can you please add specs to this. At least the `execute_as` behaviour should be solidified by specs.

[edit] sorry, haven't read @Azure7111's comment before writing.

NobodysNightmare committed:
split specs for mail notification jobs per class

NobodysNightmare committed:
tests for execute_as behaviour

NobodysNightmare committed:
please the hound

ulferts commented:
@NobodysNightmare the specs refer to shared examples which do not exist.

NobodysNightmare commented:
@ulferts Lets try again ^^

myabc commented on spec/workers/mail_notification_jobs/shared_examples.rb:
```diff
@@ -0,0 +1,100 @@
+    end
+
+    context 'raising exceptions' do
+      before { MailNotificationJob.raise_exceptions = true }
```
@NobodysNightmare try `around { |example| example.run }` instead

NobodysNightmare replied:
@myabc You are saying that:

``` ruby
before { MailNotificationJob.raise_exceptions = true }
after { MailNotificationJob.raise_exceptions = false }
```

and

``` ruby
around { |example|
  MailNotificationJob.raise_exceptions = true
  example.run
  MailNotificationJob.raise_exceptions = false
}
```

are equivalent?

This is nice to know, I'll try to remember the next time I am writing a spec of this kind. So thank you :)

However: I'd rather want this PR to be merged now. This code wouldn't be in the diff if it wasn't for the fact that I split up the specs for multiple classes into multiple files. My boyscout contribution for this spec is already done ;-)

myabc replied:
@NobodysNightmare yep, they're equivalent – it makes intention slightly clearer IMO.
